prov/tcp/src/tcpx_ep.c

```diff
@@ -0,0 +1,769 @@
+			goto err;
+		}
+	}
+
+	*ep_fid = &ep->util_ep.ep_fid;
+	(*ep_fid)->fid.ops = &tcpx_ep_fi_ops;
+	(*ep_fid)->ops = &tcpx_ep_ops;
+	(*ep_fid)->cm = &tcpx_cm_ops;
+	(*ep_fid)->msg = &tcpx_msg_ops;
+
+	return 0;
+err:
+	free(ep);
+	return ret;
+}
+
+static int tcpx_pep_fi_close(struct fid *fid)
+{
+	struct tcpx_pep *pep;
+	int ret;
+	char c;
+
+	pep = container_of(fid, struct tcpx_pep, pep.fid);
+	pep->cm.do_listen = 0;
+
+	ret = ofi_write_socket(pep->cm.signal_fds[0], &c, 1);
+	if (ret != 1)
+		FI_DBG(&tcpx_prov, FI_LOG_EP_CTRL,"Failed to signal\n");
+
+	if (pep->cm.listener_thread &&
+	    pthread_join(pep->cm.listener_thread, NULL)) {
+		FI_DBG(&tcpx_prov, FI_LOG_EP_CTRL,"pthread join failed\n");
+	}
+
+	ofi_close_socket(pep->cm.signal_fds[0]);
+	ofi_close_socket(pep->cm.signal_fds[1]);
+
+	free(pep);
+	return 0;
+
+}
+
+static int tcpx_pep_fi_bind(struct fid *fid, struct fid *bfid, uint64_t flags)
+{
+	struct tcpx_pep *pep;
+	struct util_eq *eq;
+
+	pep = container_of(fid, struct tcpx_pep, pep.fid);
+
+	if (bfid->fclass != FI_CLASS_EQ)
+		return -FI_EINVAL;
+
+	eq = container_of(bfid, struct util_eq, eq_fid.fid);
+	if (pep->fabric != eq->fabric) {
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"Cannot bind Passive EP and EQ on different fabric\n");
+		return -FI_EINVAL;
+	}
+	pep->eq = eq;
+	return 0;
+}
+
+static struct fi_ops tcpx_pep_fi_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = tcpx_pep_fi_close,
+	.bind = tcpx_pep_fi_bind,
+	.control = fi_no_control,
+	.ops_open = fi_no_ops_open,
+};
+
+static void *tcpx_cm_listener_thread(void *pep_data)
+{
+	struct tcpx_pep *pep = (struct tcpx_pep *) pep_data;
+	struct fi_eq_cm_entry *cm_entry = NULL;
+	struct tcpx_conn_handle *handle;
+	struct pollfd poll_fds[2];
+	int ret = 0, conn_fd, entry_sz;
+	char tmp = 0;
+
+	FI_DBG(&tcpx_prov, FI_LOG_EP_CTRL,
+	       "Starting listener thread for PEP: %p\n", pep);
+
+	poll_fds[0].fd = pep->cm.sock;
+	poll_fds[1].fd = pep->cm.signal_fds[1];
+	poll_fds[0].events = poll_fds[1].events = POLLIN;
+
+	while (*((volatile int *) &pep->cm.do_listen)) {
+		/* block on the listen fd until a req comes through */
+		ret = poll(poll_fds, 2, -1);
+		if (ret > 0) {
+			if (poll_fds[1].revents & POLLIN) {
+				ret = ofi_read_socket(pep->cm.signal_fds[1], &tmp, 1);
+				if (ret != 1)
+					FI_DBG(&tcpx_prov, FI_LOG_EP_CTRL,
+					       "Invalid signal\n");
+				continue;
+			}
+
+		} else {
+			break;
+		}
+
+		conn_fd = accept(pep->cm.sock, NULL, 0);
+		if (conn_fd < 0) {
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"failed to accept: %d\n", errno);
+			continue;
+		}
+
+		handle = calloc(1, sizeof(*handle));
+		if (!handle) {
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"cannot allocate memory\n");
+
+			ofi_close_socket(conn_fd);
+			break;
+		}
+		handle->conn_fd = conn_fd;
+
+		entry_sz = sizeof(*cm_entry);
+		cm_entry = calloc(1, entry_sz);
+		if (!cm_entry) {
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"cannot allocate memory\n");
+			goto err;
```

Comment:
This is unlikely to happen, but if it does, close the socket and continue processing.